{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# On Target Score Evalution"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This section references CRISPRon[1] methods and code. References for this module are listed in the References section at the end of the document.\n",
    "\n",
    "CRISPRon's codebase is: https://github.com/RTH-tools/crispron/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "##########################################################################\n",
    "#  This program is free software: you can redistribute it and/or modify\n",
    "#  it under the terms of the GNU Affero General Public License as published by\n",
    "#  the Free Software Foundation, either version 3 of the License, or\n",
    "#  (at your option) any later version.\n",
    "#\n",
    "#  It is distributed in the hope that it will be useful,\n",
    "#  but WITHOUT ANY WARRANTY; without even the implied warranty of\n",
    "#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the\n",
    "#  GNU Affero General Public License for more details.\n",
    "#\n",
    "#  You should have received a copy of the GNU Affero General Public License\n",
    "#  along with this script, see file COPYING.\n",
    "#  If not, see <http://www.gnu.org/licenses/>.\n",
    "##########################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Environment configuration and function definitions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "import glob\n",
    "from collections import OrderedDict \n",
    "import numpy as np\n",
    "os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'\n",
    "from tensorflow.keras import models\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import pickle, subprocess, gzip\n",
    "from math import exp"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Components of CRISPRspec_CRISPRoff "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "##############################################################\n",
    "#################### Energy computation #########################\n",
    "##############################################################\n",
    "\n",
    "ENERGY_MODELS_PICKLE_FILE = \"energy_dics.pkl\"\n",
    "\n",
    "RI_REV_NT_MAP = {'-':'', 'a':'T', 'A':'T', 'c':'G', 'C':'G', 'g':'C', 'G':'C',\n",
    "              't':'A', 'T':'A', 'u':'A', 'U':'A', 'n':'N', 'N':'N'}\n",
    "\n",
    "RI_DNA_DNA_NN={'AA':{'TT':-1.00}, 'TT':{'AA':-1.00}, 'AT':{'TA':-0.88}, 'TA':{'AT':-0.58},\n",
    "            'CA':{'GT':-1.45}, 'TG':{'AC':-1.45}, 'GT':{'CA':-1.44}, 'AC':{'TG':-1.44},\n",
    "            'CT':{'GA':-1.28}, 'AG':{'TC':-1.28}, 'GA':{'CT':-1.30}, 'TC':{'AG':-1.30},\n",
    "            'CG':{'GC':-2.17}, 'GC':{'CG':-2.24}, 'GG':{'CC':-1.84}, 'CC':{'GG':-1.84}}\n",
    "\n",
    "RI_MATCH_noGU = {'A':{'A':False, 'C':False, 'G':False, 'T':True},\n",
    "         'C':{'A':False, 'C':False, 'G':True, 'T':False},\n",
    "         'G':{'A':False, 'C':True, 'G':False, 'T':False},\n",
    "         'T':{'A':True, 'C':False, 'G':False, 'T':False}}\n",
    "\n",
    "########## READ THE 2mer 3mer 4mer energies ###############\n",
    "RNA_DNA_internal_loop ={3:3.2, 4:3.555, 5:3.725, 6:3.975, 7:4.16, 8:4.33, 9:4.495, 10:4.6, 11:4.7}\n",
    "RNA_DNA = None\n",
    "def read_energy_parameters(ENERGY_MODELS_PICKLE_FILE = \"energy_dics.pkl\"):\n",
    "    global RNA_DNA\n",
    "    energy_reader=open(ENERGY_MODELS_PICKLE_FILE, 'rb')\n",
    "    RNA_DNA = pickle.load(energy_reader)\n",
    "    energy_reader.close()\n",
    "\n",
    "####### Necessary for self-folding ######\n",
    "RNAFOLD_EXE = \"RNAfold\"\n",
    "\n",
    "######### positional energy contribution weights ###################\n",
    "# LAST weight is filled but not used, (Left-over from some experimental option)\n",
    "POS_WGH=[1.80067099242007, 1.95666668400006, 1.90472004401173, 2.13047270152512, 1.37853848098249, 1.46460783730408, 1.0, 1.387220146823, 1.51401000729362, 1.98058344620751, 1.87939168587699, 1.7222593588838, 2.02228445489326, 1.92692086621503, 2.08041972716723, 1.94496755678903, 2.14539112893591, 2.04277109036766, 2.24911493451185, 2.25]\n",
    "\n",
    "# LAST weight is filled but not used, (Left-over from some experimental option)\n",
    "DNA_POS_WGH = [1.22245576981774, 1.24561578622024, 1.37883177517399, 1.39146340276523, 1.24308180746857, 1.09598194424544, 1.0, 1.11695025382169, 1.11589045394936, 1.22243614188218, 1.21317477033274, 1.07125942316357, 1.25205871414019, 1.21445408158483, 1.20971491326295, 1.21076785001579, 1.2480898972246, 1.40301355270318, 1.41221084925493, 1.4]\n",
    "\n",
    "######### pam correction parameters for pam-updated energy ##########\n",
    "pam_ratios={\"GGG\":1.0, \"AGG\":1.0, \"CGG\":1.0, \"TGG\":1.0, \"GAG\":0.9, \"AAG\":0.9, \"CAG\":0.9, \"TAG\":0.9, \"GGA\":0.8, \"AGA\":0.8, \"CGA\":0.8, \"TGA\":0.8, \"OTHERS\": 0.0}\n",
    "pam_ratio_count = 3\n",
    "\n",
    "########## gRNA folding #######################\n",
    "grna_folding_engs = {}\n",
    "def get_rnafold_eng(seq, rid=\"temp_grna_id\"):\n",
    "    if seq not in grna_folding_engs:\n",
    "        no_constraint_eng = None\n",
    "        l = len(seq)\n",
    "        cmd = RNAFOLD_EXE\n",
    "        p = subprocess.Popen([cmd, '--noPS'], shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True).communicate(input=\">\"+rid+\"\\n\"+seq+\"\\n\\n\")\n",
    "        if p[1]==\"\":\n",
    "            no_constraint_eng = float(p[0].rstrip().split()[-1].replace('(','').replace(')',''))\n",
    "        else:\n",
    "            sys.stderr.write(\"#ERROR:RNAfold run went wrong: \"+cmd+\"; \"+p[1]+\"\\n\")\n",
    "            exit()\n",
    "        grna_folding_engs[seq] = no_constraint_eng\n",
    "    return grna_folding_engs[seq]\n",
    "\n",
    "# Stacking energy is distributed to positions only for interior loops\n",
    "def calcRNADNAenergy(guideSeq, otSeq, GU_allowed=False):\n",
    "    guideSeq = guideSeq.upper()[:-3]\n",
    "    seq = ''.join([RI_REV_NT_MAP[c] for c in otSeq[:-3]])\n",
    "\n",
    "    spos = -1\n",
    "    epos = -1\n",
    "\n",
    "    energy = [0.0]*len(guideSeq)\n",
    "\n",
    "    MATCH = RI_MATCH_noGU\n",
    "\n",
    "    for i in range(len(seq)):\n",
    "        if MATCH[guideSeq[i]][seq[i]]:\n",
    "            if spos==-1:\n",
    "                spos = i\n",
    "            epos = i\n",
    "\n",
    "    i = spos\n",
    "    while i<epos:\n",
    "        j = i+1\n",
    "        while MATCH[seq[j]][guideSeq[j]]==False:\n",
    "            j = j+1\n",
    "\n",
    "            if j > epos:\n",
    "                break\n",
    "        if j > epos:\n",
    "            break\n",
    "\n",
    "        loop_size = (j-i)-1\n",
    "        eng_con = 0\n",
    "        if loop_size<3:\n",
    "            eng_con = RNA_DNA[loop_size][guideSeq[i:j+1]][seq[i:j+1]]\n",
    "            # if there is a stack in the beginning or end AU GU penalty is still needed\n",
    "            if loop_size == 0:\n",
    "                if (i==spos and (guideSeq[i]==\"T\" or seq[i]==\"T\")) or (j==epos and (guideSeq[j]==\"T\" or seq[j]==\"T\")):\n",
    "                    eng_con += 0.25\n",
    "        else:\n",
    "            eng_con = float(RNA_DNA_internal_loop[loop_size]) + float(RNA_DNA[0][guideSeq[i:i+2]][seq[i:i+2]]) + float(RNA_DNA[0][guideSeq[j-1:j+1]][seq[j-1:j+1]])\n",
    "\n",
    "        for k in range(loop_size+1):\n",
    "            energy[i+k] += eng_con/(loop_size+1)\n",
    "\n",
    "        i = j\n",
    "\n",
    "    return energy\n",
    "\n",
    "################ DNA-DNA opening #############\n",
    "def calcDNAopeningScore(otSeq):\n",
    "    seq = otSeq.upper()[:-3]\n",
    "    energy = [0.0]*len(seq)\n",
    "    for i in range(1, len(seq)):\n",
    "        energy[i] = float(RI_DNA_DNA_NN[seq[i-1]+seq[i]][RI_REV_NT_MAP[seq[i-1]]+RI_REV_NT_MAP[seq[i]]])\n",
    "    return energy\n",
    "#################################################################################################\n",
    "\n",
    "REV_NT_MAP = {'-':'', 'a':'T', 'A':'T', 'c':'G', 'C':'G', 'g':'C', 'G':'C', \n",
    "              't':'A', 'T':'A', 'u':'A', 'U':'A', 'n':'N', 'N':'N'}\n",
    "\n",
    "def rev_comp_seq(seq):\n",
    "    s=''\n",
    "    for c in seq:\n",
    "        s =  REV_NT_MAP[c] + s\n",
    "    return s\n",
    "\n",
    "def comp_seq(seq):\n",
    "    s=''\n",
    "    for c in seq:\n",
    "        s =  s + REV_NT_MAP[c]\n",
    "    return s\n",
    "\n",
    "############# Get interaction energy ##################\n",
    "#Employ all the necessary computations on the score vector to get the final free energy\n",
    "def get_eng(grna_seq, off_seq, score_func, GU_allowed=False, pos_weight=False, pam_corr=False, grna_folding=False, dna_opening=False, dna_pos_wgh=False):\n",
    "    scores = score_func(grna_seq, off_seq, GU_allowed)\n",
    "\n",
    "    if pos_weight:\n",
    "        for i in range(len(scores)):\n",
    "            if i < 21:\n",
    "                scores[-(i+1)] = POS_WGH[-(i+1)] * scores[-(i+1)]\n",
    "\n",
    "    off = sum(scores) + 0.0\n",
    "    off = (-1.0) * off\n",
    "\n",
    "    if grna_folding:\n",
    "        off += get_rnafold_eng(grna_seq[:20])\n",
    "\n",
    "\n",
    "    if dna_opening:\n",
    "        dna_scores = calcDNAopeningScore(off_seq)\n",
    "        if dna_pos_wgh:\n",
    "            for i in range(len(dna_scores)):\n",
    "                if i < 21:\n",
    "                    dna_scores[-(i+1)] = DNA_POS_WGH[-(i+1)] * dna_scores[-(i+1)]\n",
    "        off += sum(dna_scores)\n",
    "\n",
    "    if pam_corr:\n",
    "        if off_seq[-pam_ratio_count:] in pam_ratios.keys():\n",
    "            off = off * pam_ratios[off_seq[-pam_ratio_count:]]\n",
    "        else:\n",
    "            off = off * pam_ratios[\"OTHERS\"]\n",
    "\n",
    "    return off\n",
    "\n",
    "#################################################################\n",
    "# TODO: Turn the ontarget and offSeq into class for ease of understanding the code #\n",
    "\n",
    "## Compute POFF for the given grna and its off-targets ##\n",
    "PAR_BETA = 1.000000 / (0.001987 * 310.150000)\n",
    "def compute_CRISPRspec(ontarget, offSeqs, score_func, GU_allowed=False, pos_weight=False, pam_corr=False, grna_folding=False, dna_opening=False, dna_pos_wgh=False, ignored_chromosomes=set()):\n",
    "    pf = 0.000000\n",
    "    on = 0.000000\n",
    "    CRISPRoff_scores = []\n",
    "\n",
    "    for offSeq in offSeqs:\n",
    "        offSeq_eng = get_eng(ontarget[0], offSeq[0], score_func,  GU_allowed=GU_allowed, pos_weight=pos_weight, pam_corr=pam_corr, grna_folding=grna_folding, dna_opening=dna_opening, dna_pos_wgh=dna_pos_wgh)\n",
    "        if offSeq[1] not in ignored_chromosomes:\n",
    "            pf = pf + exp(PAR_BETA * offSeq_eng)\n",
    "        else:\n",
    "            sys.stderr.write(\"#WARNING: This off-target sequence is ignored when computing CRISPRspec: '\"+\"|\".join([str(x) for x in offSeq])+\"'.\\n\")\n",
    "        CRISPRoff_scores.append((offSeq, offSeq_eng))\n",
    "\n",
    "    on_eng = get_eng(ontarget[0], ontarget[0], score_func, GU_allowed=GU_allowed, pos_weight=pos_weight, pam_corr=pam_corr, grna_folding=grna_folding, dna_opening=dna_opening, dna_pos_wgh=dna_pos_wgh)\n",
    "    on = exp(PAR_BETA * on_eng)\n",
    "\n",
    "    CRISPRoff_scores.append(((ontarget[0], ontarget[2], ontarget[3], ontarget[4], ontarget[5]), on_eng))\n",
    "    return (pf/(pf+on)), CRISPRoff_scores\n",
    "\n",
    "###############################################################\n",
    "################ Input readers ################################\n",
    "###############################################################\n",
    "# Returns all valid Cas9 targets predicted with RIsearch2 in the genome\n",
    "# Result file must have been generated with -p3 option\n",
    "# grna is the sequence without the PAM addition.\n",
    "# If PAM is added last parameter must be set to False\n",
    "def read_risearch_results(guideSeq, ris_file, noPAM_given=True, count_mms=False, on_targets = [], offSeqs = [], off_counts = {\"GG\":[0]*7, \"AG\":[0]*7, \"GA\":[0]*7}):\n",
    "    inf = gzip.open(ris_file, 'rt') if ris_file.endswith(\".gz\") else open(ris_file, 'rt')\n",
    "    x = 0\n",
    "    for line in inf:\n",
    "        x = x + 1\n",
    "        cols = line.rstrip().split(\"\\t\")\n",
    "        if len(cols)>11 and len(cols[10])>3 and len(cols[11])>0:\n",
    "            gid, qs, qe, tc, ts, te, tst, en, ist, iseq, pamseq, preseq= cols[:12]\n",
    "            PAM = comp_seq(pamseq[:3])\n",
    "            if PAM[1:3] in [\"GG\", \"AG\", \"GA\"]:\n",
    "                offseq = comp_seq(iseq)\n",
    "                if tst==\"+\":\n",
    "                    ts = str(int(ts)-4)\n",
    "                    tst=\"-\"\n",
    "                else:\n",
    "                    ts = str(int(ts)-1)\n",
    "                    te = str(int(te)+3)\n",
    "                    tst=\"+\"\n",
    "\n",
    "                # determine and save the number of mismatches for this guide\n",
    "                mm_count = sum ( offseq[i] != guideSeq[i] for i in range(len(offseq)) )\n",
    "                if count_mms:\n",
    "                    if mm_count<7:\n",
    "                        off_counts[PAM[1:3]][mm_count] += 1\n",
    "                if mm_count<7:\n",
    "                    if (noPAM_given and (guideSeq != offseq)) or ((noPAM_given==False) and (guideSeq != offseq+PAM)):\n",
    "                        offSeqs.append(((offseq+PAM), tc, ts, te, tst))\n",
    "                    else:\n",
    "                        on_targets.append(((offseq+PAM), (rev_comp_seq(preseq[:4])+offseq+comp_seq(pamseq[:6])), tc, ts, te, tst))\n",
    "        else:\n",
    "            sys.stderr.write(\"WARNING: RIsearch output line is missing columns, or info in columns.\\nLINE: \"+line)\n",
    "    inf.close()\n",
    "    return  offSeqs, off_counts, on_targets\n",
    "\n",
    "# Returns all valid Cas9 targets from Cas-OFFinder result file\n",
    "def read_casoff_results(guideSeq, casoff_file, count_mms=False):\n",
    "    on_targets = []\n",
    "    offSeqs = []\n",
    "    off_counts = {\"GG\":[0]*7, \"AG\":[0]*7, \"GA\":[0]*7}\n",
    "    inf = gzip.open(casoff_file, 'rt') if casoff_file.endswith(\".gz\") else open(casoff_file, 'rt')\n",
    "    x = 0\n",
    "    for line in inf:\n",
    "        x = x + 1\n",
    "        cols = line.rstrip().split(\"\\t\")\n",
    "        if line[0]!=\"#\" and len(cols)>7 and cols[0]==\"X\":\n",
    "            bulge_type, crRNA, DNA, chromosome, position, strand, mismatches, bulge_size = cols[:8]\n",
    "            PAM = DNA.upper()[-3:]\n",
    "            if PAM[1:3] in [\"GG\", \"AG\", \"GA\"]:\n",
    "                offseq = DNA.upper()\n",
    "                mm_count = int(mismatches)\n",
    "                # determine and save the number of mismatches for this guide\n",
    "                if count_mms:\n",
    "                    if mm_count<7:\n",
    "                        off_counts[PAM[1:3]][mm_count] += 1\n",
    "\n",
    "                if mm_count<7:\n",
    "                    if DNA[:len(guideSeq)]==guideSeq:\n",
    "                        if strand==\"+\":\n",
    "                            on_targets.append(((offseq), DNA, chromosome, position, str(int(position)+len(offseq)), strand))\n",
    "                        else:\n",
    "                            on_targets.append(((offseq), DNA, chromosome, str(int(position)-len(offseq)), position, strand))\n",
    "                    else:\n",
    "                        if strand==\"+\":\n",
    "                            offSeqs.append(((offseq), chromosome, position, str(int(position)+len(offseq)), strand))\n",
    "                        else:\n",
    "                            offSeqs.append(((offseq), chromosome, str(int(position)-len(offseq)), position, strand))\n",
    "    inf.close()\n",
    "    return  offSeqs, off_counts, on_targets\n",
    "\n",
    "# Returns all valid Cas9 targets from plain new-line seperated input off-targets file\n",
    "def read_standard_offtargets_input(guideSeq, in_file, count_mms=False):\n",
    "    on_targets = []\n",
    "    offSeqs = []\n",
    "    off_counts = {\"GG\":[0]*7, \"AG\":[0]*7, \"GA\":[0]*7}\n",
    "    inf = gzip.open(in_file, 'rt') if in_file.endswith(\".gz\") else open(in_file, 'rt')\n",
    "    x = 0\n",
    "    for line in inf:\n",
    "        cols = line.rstrip().split(\"\\t\")\n",
    "        if line[0]!=\"#\" and len(cols[0])==len(guideSeq):\n",
    "            x = x + 1\n",
    "            offseq = cols[0].upper()\n",
    "            PAM = offseq[-3:]\n",
    "            if PAM[1:3] in [\"GG\", \"AG\", \"GA\"]:\n",
    "                mm_count = sum ( offseq[i] != guideSeq[i] for i in range(len(offseq)-3) )\n",
    "                # determine and save the number of mismatches for this guide\n",
    "                if count_mms:\n",
    "                    if mm_count<7:\n",
    "                        off_counts[PAM[1:3]][mm_count] += 1\n",
    "                if mm_count<7:\n",
    "                    if offseq==guideSeq:\n",
    "                        on_targets.append((offseq, offseq, None, None, None, None))\n",
    "                    else:\n",
    "                        offSeqs.append((offseq, None, None, None, None))\n",
    "    inf.close()\n",
    "    return  offSeqs, off_counts, on_targets\n",
    "\n",
    "#Read given off-target file whether its risearch casoff or standard regular new-line seperated file\n",
    "def read_offtargets_file(guideSeq, offtargets_file, noPAM_given=False, count_mms=False, chromosome_names=None):\n",
    "    if \"risearch\" in offtargets_file:\n",
    "        if chromosome_names==None:\n",
    "            sys.stdout.write('#RUNNING: reading given risearch output file \"'+offtargets_file+'\" for gRNA/on-target:'+guideSeq+'.\\n')\n",
    "            return read_risearch_results(guideSeq, offtargets_file, noPAM_given=noPAM_given, count_mms=count_mms, on_targets = [], offSeqs = [], off_counts = {\"GG\":[0]*7,\"AG\":[0]*7,\"GA\":[0]*7})\n",
    "        else: # READ from multiple files\n",
    "            chr_inf = gzip.open(chromosome_names, 'rt') if chromosome_names.endswith(\".gz\") else open(chromosome_names, 'rt')\n",
    "            on_targets = []\n",
    "            offSeqs = []\n",
    "            off_counts = {\"GG\":[0]*7, \"AG\":[0]*7, \"GA\":[0]*7}\n",
    "            ris_dir = \"/\".join(offtargets_file.split(\"/\")[:-1])\n",
    "            ris_file = offtargets_file.split(\"/\")[-1]\n",
    "            for chrid in chr_inf:\n",
    "                sys.stdout.write('#RUNNING: reading given risearch output file \"'+\"/\".join([ris_dir, chrid.rstrip(), ris_file])+'\".\\n')\n",
    "                if os.path.isfile(\"/\".join([ris_dir, chrid.rstrip(), ris_file])):\n",
    "                    offSeqs, off_counts, on_targets = read_risearch_results(guideSeq, \"/\".join([ris_dir, chrid.rstrip(), ris_file]), noPAM_given=noPAM_given, count_mms=count_mms, on_targets = on_targets, offSeqs = offSeqs, off_counts = off_counts)\n",
    "            chr_inf.close()\n",
    "            return offSeqs, off_counts, on_targets\n",
    "    else:\n",
    "        inf = gzip.open(offtargets_file, 'rt') if offtargets_file.endswith(\".gz\") else open(offtargets_file, 'rt')\n",
    "        casoff = True if inf.readline().startswith(\"#Bulge\") else False\n",
    "        inf.close()\n",
    "        if casoff:\n",
    "            sys.stdout.write('#RUNNING: reading given Cas-OFFinder file \"'+offtargets_file+'\".\\n')\n",
    "            return read_casoff_results(guideSeq, offtargets_file, count_mms=count_mms)\n",
    "        else:\n",
    "            sys.stdout.write('#RUNNING: reading given offtarget file \"'+offtargets_file+'\" as regular off-target file.\\n')\n",
    "            return read_standard_offtargets_input(guideSeq, offtargets_file, count_mms=count_mms)\n",
    "\n",
    "# Read guide sequences from fasta file\n",
    "def read_guides_fasta(fasta_file):\n",
    "    ontargets={}\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        seq = str(record.seq).upper()\n",
    "        if len(str(record.seq))==23:\n",
    "            ontargets[record.id] = seq\n",
    "        else:\n",
    "            # get valid guides with PAM NGG and 23nt\n",
    "            x=0\n",
    "            grna_seq=\"\"\n",
    "            for i in range(len(seq)):\n",
    "                if seq[i] not in [\"A\", \"C\", \"G\", \"T\", \"U\"]:\n",
    "                    grna_seq =\"\"\n",
    "                elif len(grna_seq)<22:\n",
    "                    grna_seq = grna_seq + seq[i]\n",
    "                else:\n",
    "                    grna_seq = grna_seq + seq[i]\n",
    "                    if grna_seq[-2:] == \"GG\":\n",
    "                        x+=1\n",
    "                        ontargets[record.id+\"_gRNA_\"+str(x)] = grna_seq\n",
    "                    grna_seq = grna_seq[1:]\n",
    "            revseq = rev_comp_seq(seq)\n",
    "            for i in range(len(revseq)):\n",
    "                if revseq[i] not in [\"A\", \"C\", \"G\", \"T\", \"U\"]:\n",
    "                    grna_seq =\"\"\n",
    "                elif len(grna_seq)<22:\n",
    "                    grna_seq = grna_seq + revseq[i]\n",
    "                else:\n",
    "                    grna_seq = grna_seq + revseq[i]\n",
    "                    if grna_seq[-2:] == \"GG\":\n",
    "                        x+=1\n",
    "                        ontargets[record.id+\"_rev_gRNA_\"+str(x)] = grna_seq\n",
    "                    grna_seq = grna_seq[1:]\n",
    "\n",
    "            if x==1:\n",
    "                ontargets[record.id] = ontargets[record.id+\"_gRNA_\"+str(x)]\n",
    "                del ontargets[record.id+\"_gRNA_\"+str(x)]\n",
    "    return ontargets\n",
    "\n",
    "## Summarize the features for given on-targets\n",
    "def get_energy_features_for_guides(guideSeqs):\n",
    "    feature_out = {}\n",
    "    for guide, guideSeq in guideSeqs.items():\n",
    "        feature_out[guide]={}\n",
    "        feature_out[guide][\"RNA_DNA_eng\"] = -1.00000 * get_eng(guideSeq, guideSeq, calcRNADNAenergy, GU_allowed=False, pos_weight=False, pam_corr=False, grna_folding=False, dna_opening=False, dna_pos_wgh=False)\n",
    "        feature_out[guide][\"RNA_DNA_eng_weighted\"] = -1.00000 * get_eng(guideSeq, guideSeq, calcRNADNAenergy, GU_allowed=False, pos_weight=True, pam_corr=False, grna_folding=False, dna_opening=False, dna_pos_wgh=False)\n",
    "        feature_out[guide][\"DNA_DNA_opening\"] = -1.00000 * sum(calcDNAopeningScore(guideSeq))\n",
    "        feature_out[guide][\"spacer_self_fold\"] = get_rnafold_eng(guideSeq[:20])\n",
    "        feature_out[guide][\"CRISPRoff_score\"] = get_eng(guideSeq, guideSeq, calcRNADNAenergy, GU_allowed=False, pos_weight=True, pam_corr=True, grna_folding=True, dna_opening=True, dna_pos_wgh=False)\n",
    "    features=[\"RNA_DNA_eng\", \"RNA_DNA_eng_weighted\", \"DNA_DNA_opening\", \"spacer_self_fold\", \"CRISPRoff_score\"]\n",
    "    return features, feature_out\n",
    "\n",
    "## Summarize the on off target info for given on_targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 如何从30mers生成23mers?\n",
    "PRE_GUIDE=4\n",
    "GUIDE=20\n",
    "PRE_PAM=1\n",
    "PAM='GG'\n",
    "nPAM=len(PAM)\n",
    "POST_PAM=3\n",
    "TOTAL=PRE_GUIDE+GUIDE+PRE_PAM+nPAM+POST_PAM\n",
    "PAMoffset=PRE_GUIDE+GUIDE+PRE_PAM\n",
    "#23mers = v[1][PRE_GUIDE:TOTAL-POST_PAM]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_30mers_fasta(fn):\n",
    "    fasta = SeqIO.parse(fn, \"fasta\")\n",
    "    d = OrderedDict()\n",
    "    for r in fasta:\n",
    "        id, seq = r.id, r.seq\n",
    "        if id in d:\n",
    "            print('sequence identifiers most be unique', file=sys.stderr)\n",
    "            sys.exit(1)\n",
    "        elif not len(seq) == eLENGTH:\n",
    "            print('\"%s\" is not a sequence of length %d' % (seq, eLENGTH), file=sys.stderr)\n",
    "            sys.exit(1)\n",
    "        else:\n",
    "            d[id] = list()\n",
    "            d[id].append(seq)\n",
    "    return d\n",
    "\n",
    "def get_23_mers(d):\n",
    "    d_23mers  = {}\n",
    "    for k, v in d.items():\n",
    "        d_23mers[k] = str(v[0][PRE_GUIDE:TOTAL-POST_PAM])\n",
    "    return d_23mers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Components of CRISPRon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(threshold=np.inf)\n",
    "\n",
    "#length of input seq\n",
    "eLENGTH=30\n",
    "#depth of onehot encoding\n",
    "eDEPTH=4\n",
    "\n",
    "def onehot(x):\n",
    "    z = list()\n",
    "    for y in list(x):\n",
    "        if y in \"Aa\":  z.append(0)\n",
    "        elif y in \"Cc\": z.append(1)\n",
    "        elif y in \"Gg\": z.append(2)\n",
    "        elif y in \"TtUu\": z.append(3)\n",
    "        else:\n",
    "            print(\"Non-ATGCU character \" + data[l], file=sys.stderr)\n",
    "            raise Exception\n",
    "    return z\n",
    "\n",
    "def set_data(DX, s):\n",
    "    for j,x in enumerate(onehot(s)):\n",
    "        DX[j][x] = 1\n",
    "\n",
    "def preprocess_seq(data):\n",
    "    dkeys = data.keys()\n",
    "    DATA_X = np.zeros((len(dkeys),eLENGTH,eDEPTH), dtype=np.float32) # onehot\n",
    "    DATA_G = np.zeros((len(dkeys)), dtype=np.float32) #deltaGb\n",
    "\n",
    "    seqs = list()\n",
    "\n",
    "    for l, id in enumerate(dkeys):\n",
    "        d = data[id]\n",
    "        set_data(DATA_X[l], d[1])\n",
    "        DATA_G[l] = d[0]\n",
    "        seqs.append(d[1])\n",
    "\n",
    "    return (seqs, DATA_X, DATA_G,)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# main function of get_deltaG()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Only need deltaG\n",
    "args_duplex_energy_params = \"../data/model/energy_dics.pkl\"\n",
    "read_energy_parameters(args_duplex_energy_params)\n",
    "def CRISPRspec_get_deltaG(guideSeqs,test_mode=False):\n",
    "    # Pre_STEP 1: Setting arguments and energy parameters.    \n",
    "    args_rnafold_x = 'RNAfold'\n",
    "    if test_mode:\n",
    "        sys.stdout.write('#STEP 1: arguments and energy parameters are ready.\\n')\n",
    "    global RNAFOLD_EXE\n",
    "    RNAFOLD_EXE = args_rnafold_x\n",
    "\n",
    "    # Pre_STEP 2: Check input guide sequences.\n",
    "    if len(guideSeqs.keys())==0:\n",
    "        sys.stdout.write('#ABORTED: No guide sequence given.\\n')\n",
    "        sys.stdout.write('#ERROR: There are no guide sequences given as input.\\n')\n",
    "        exit()\n",
    "\n",
    "    # STEP 3: Get on-target energy features \n",
    "    feature_names, feature_values_dic = get_energy_features_for_guides(guideSeqs)\n",
    "    deltaG_result=pd.DataFrame(columns=['guideID', 'guideSeq']+feature_names)\n",
    "    for guide, feature_dic in feature_values_dic.items():\n",
    "            tmp=[guide, guideSeqs[guide]]+[str(feature_dic[feature]) for feature in feature_names]\n",
    "            #deltaG_result = deltaG_result.append(tmp,ignore_index= True)\n",
    "            deltaG_result.loc[len(deltaG_result)] = tmp\n",
    "\n",
    "    # FINISH\n",
    "    if test_mode:\n",
    "        sys.stdout.write('#END: FINISHED RUNNING THE PIPELINE WITH NO ERROR.\\n')\n",
    "        sys.stdout.write('#CRISPRspec_pipeline finished.\\n')\n",
    "\n",
    "    return deltaG_result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# main function of get_on_score()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_on_score(d,MODELD):\n",
    "    # id, deltaGb, seq\n",
    "    (s, x, g) = preprocess_seq(d)\n",
    "    tinput = list()\n",
    "    tinput.append(x)\n",
    "    tinput.append(g)\n",
    "\n",
    "    #evaluate per model\n",
    "    v = list()\n",
    "    for i, md in enumerate(MODELD):\n",
    "        model = models.load_model(md)\n",
    "        model.compile(loss='mse', metrics=['mae', 'mse'])\n",
    "        predicted = model.predict(tinput, use_multiprocessing=True, workers=32)\n",
    "        predicted = [p[0] for p in predicted]\n",
    "        v.append([list(d.keys()), s, predicted])\n",
    "\n",
    "    #average model outputs\n",
    "    ids = v[0][0]\n",
    "    sequences = v[0][1]\n",
    "    predicted = [0 for i in range(len(sequences))]\n",
    "    nmodels = len(MODELD)\n",
    "    for m in range(nmodels):\n",
    "        predicted = [predicted[i] + v[m][2][i]/nmodels for i in range(len(sequences))]\n",
    "\n",
    "    result=pd.DataFrame({'ID':ids,'30mer_sequences':sequences,'On_Scores':predicted})\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def seqs_add_deltaG(d, crisproff_df):    \n",
    "    for index, row in crisproff_df.iterrows():\n",
    "        id, g = row[0], row[6]\n",
    "        try:\n",
    "            g = float(g)\n",
    "        except:\n",
    "            sys.print(\"ERROR: deltaG value is not float: \"+g+\"\\n\")\n",
    "            sys.exit(1)\n",
    "        d[id].insert(0, g)\n",
    "    return d"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Execution:Predict On_target score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def CRISPROn_score_pipeline(d):\n",
    "    #For compute CRISPRon scores\n",
    "    MODELD=glob.glob('../data/deep_models/best/*/') #paths to model(s) #$DATADIR/deep_models/best/*/\n",
    "\n",
    "    #Get seqs: 30mers-> 23mers\n",
    "    guideSeqs = get_23_mers(d)\n",
    "\n",
    "    #Get deltaG\n",
    "    deltaG_result= CRISPRspec_get_deltaG(guideSeqs=guideSeqs,test_mode=False)\n",
    "    data = seqs_add_deltaG(d, deltaG_result)\n",
    "    result=get_on_score(data,MODELD)\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = '../test/outdir.test/30mers.fa'\n",
    "seq_data = read_30mers_fasta(fn=fasta_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:SavedModel saved prior to TF 2.5 detected when loading Keras model. Please ensure that you are saving the model with model.save() or tf.keras.models.save_model(), *NOT* tf.saved_model.save(). To confirm, there should be a file named \"keras_metadata.pb\" in the SavedModel directory.\n",
      "11/11 [==============================] - 3s 203ms/step\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>30mer_sequences</th>\n",
       "      <th>On_Scores</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>test_seq_p_19</td>\n",
       "      <td>(T, G, C, C, A, T, G, C, A, A, G, G, C, A, A, ...</td>\n",
       "      <td>50.426987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>test_seq_p_22</td>\n",
       "      <td>(C, A, T, G, C, A, A, G, G, C, A, A, A, C, T, ...</td>\n",
       "      <td>49.218208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>test_seq_p_23</td>\n",
       "      <td>(A, T, G, C, A, A, G, G, C, A, A, A, C, T, G, ...</td>\n",
       "      <td>49.222065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>test_seq_p_24</td>\n",
       "      <td>(T, G, C, A, A, G, G, C, A, A, A, C, T, G, C, ...</td>\n",
       "      <td>60.418915</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>test_seq_p_25</td>\n",
       "      <td>(G, C, A, A, G, G, C, A, A, A, C, T, G, C, T, ...</td>\n",
       "      <td>53.734760</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>336</th>\n",
       "      <td>test_seq2_m_62</td>\n",
       "      <td>(C, A, G, T, T, C, T, G, T, G, C, T, T, C, C, ...</td>\n",
       "      <td>13.416720</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>337</th>\n",
       "      <td>test_seq2_m_28</td>\n",
       "      <td>(G, T, G, A, C, T, A, T, T, G, T, A, C, T, T, ...</td>\n",
       "      <td>33.848930</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>338</th>\n",
       "      <td>test_seq2_m_9</td>\n",
       "      <td>(T, G, T, A, G, A, G, G, A, G, A, G, A, A, A, ...</td>\n",
       "      <td>58.765717</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>339</th>\n",
       "      <td>test_seq2_m_8</td>\n",
       "      <td>(G, T, A, G, A, G, G, A, G, A, G, A, A, A, A, ...</td>\n",
       "      <td>43.916271</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>340</th>\n",
       "      <td>test_seq2_m_7</td>\n",
       "      <td>(T, A, G, A, G, G, A, G, A, G, A, A, A, A, A, ...</td>\n",
       "      <td>63.997688</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>341 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 ID                                    30mer_sequences  \\\n",
       "0     test_seq_p_19  (T, G, C, C, A, T, G, C, A, A, G, G, C, A, A, ...   \n",
       "1     test_seq_p_22  (C, A, T, G, C, A, A, G, G, C, A, A, A, C, T, ...   \n",
       "2     test_seq_p_23  (A, T, G, C, A, A, G, G, C, A, A, A, C, T, G, ...   \n",
       "3     test_seq_p_24  (T, G, C, A, A, G, G, C, A, A, A, C, T, G, C, ...   \n",
       "4     test_seq_p_25  (G, C, A, A, G, G, C, A, A, A, C, T, G, C, T, ...   \n",
       "..              ...                                                ...   \n",
       "336  test_seq2_m_62  (C, A, G, T, T, C, T, G, T, G, C, T, T, C, C, ...   \n",
       "337  test_seq2_m_28  (G, T, G, A, C, T, A, T, T, G, T, A, C, T, T, ...   \n",
       "338   test_seq2_m_9  (T, G, T, A, G, A, G, G, A, G, A, G, A, A, A, ...   \n",
       "339   test_seq2_m_8  (G, T, A, G, A, G, G, A, G, A, G, A, A, A, A, ...   \n",
       "340   test_seq2_m_7  (T, A, G, A, G, G, A, G, A, G, A, A, A, A, A, ...   \n",
       "\n",
       "     On_Scores  \n",
       "0    50.426987  \n",
       "1    49.218208  \n",
       "2    49.222065  \n",
       "3    60.418915  \n",
       "4    53.734760  \n",
       "..         ...  \n",
       "336  13.416720  \n",
       "337  33.848930  \n",
       "338  58.765717  \n",
       "339  43.916271  \n",
       "340  63.997688  \n",
       "\n",
       "[341 rows x 3 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result=CRISPROn_score_pipeline(seq_data)\n",
    "result"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read N20s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "n20_target=pd.read_excel('./n20.xlsx',header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna, generic_rna\n",
    "from collections import OrderedDict\n",
    "\n",
    "# Complement and transcribe\n",
    "n20_target['RNA'] = n20_target.iloc[:,0].apply(lambda x: str(Seq(x)))\n",
    "seq_data_1 =  n20_target['RNA']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_data_2 = seq_data_1.apply(lambda x: 'AGAT' + x + 'GGGTTA')\n",
    "seq_data_1 = seq_data_1.apply(lambda x: 'ACAA' + x + 'GGGTTA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert2Dict(seq_data_Series):\n",
    "    # Convert to Seq type dict\n",
    "    tmp = OrderedDict()\n",
    "    for i, seq in enumerate(seq_data_Series):\n",
    "        seq_id = 'test_seq_{}'.format(i+1)\n",
    "        tmp[seq_id] = list()\n",
    "        tmp[seq_id].append(Seq(seq))\n",
    "        \n",
    "    seq_data_Series=tmp\n",
    "    del tmp\n",
    "    return seq_data_Series\n",
    "seq_data_1=convert2Dict(seq_data_1)\n",
    "seq_data_2=convert2Dict(seq_data_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "OrderedDict([('test_seq_1', [Seq('ACAAATGTAACTAGTTAGTGAGACGGGTTA')]),\n",
       "             ('test_seq_2', [Seq('ACAATTAGTGAGACCCTAGGTGTAGGGTTA')]),\n",
       "             ('test_seq_3', [Seq('ACAACCTAGTTAGTAGCTACCTAGGGGTTA')]),\n",
       "             ('test_seq_4', [Seq('ACAAGTCCTTCTAGCCTAGTTAGTGGGTTA')]),\n",
       "             ('test_seq_5', [Seq('ACAACCTAGGTGTATCTAAAGTCTGGGTTA')]),\n",
       "             ('test_seq_6', [Seq('ACAACTAGTTTAGAGACTCACGAGGGGTTA')]),\n",
       "             ('test_seq_7', [Seq('ACAAACCTAAGTATTTAGTGAGACGGGTTA')]),\n",
       "             ('test_seq_8', [Seq('ACAAAGCTACCTAGGACTCACGAGGGGTTA')]),\n",
       "             ('test_seq_9', [Seq('ACAAGACCCCCCTCCTAGTTTAGAGGGTTA')]),\n",
       "             ('test_seq_10', [Seq('ACAACCTAGGACCCCTAGTTTAGAGGGTTA')]),\n",
       "             ('test_seq_11', [Seq('ACAAGACCCCCCTCGACCCCCCTCGGGTTA')]),\n",
       "             ('test_seq_12', [Seq('ACAAGTCTATAGTAGTCTATAGTAGGGTTA')]),\n",
       "             ('test_seq_13', [Seq('ACAACCTCCCTACTTCTAAAGTCTGGGTTA')]),\n",
       "             ('test_seq_14', [Seq('ACAACCTCCCTACTACCTAAGTATGGGTTA')]),\n",
       "             ('test_seq_15', [Seq('ACAAGACCCCCCTCAGCTACCTAGGGGTTA')]),\n",
       "             ('test_seq_16', [Seq('ACAATAGACCTAGACCTAGTTAGTGGGTTA')]),\n",
       "             ('test_seq_17', [Seq('ACAATCTAAAGTCTCAAGAAGTATGGGTTA')]),\n",
       "             ('test_seq_18', [Seq('ACAATCTAAAGTCTCCTAGGACCCGGGTTA')]),\n",
       "             ('test_seq_19', [Seq('ACAATCTAAAGTCTGTCCTTCTAGGGGTTA')]),\n",
       "             ('test_seq_20', [Seq('ACAACAAGAAGTATGACCCCCCTCGGGTTA')]),\n",
       "             ('test_seq_21', [Seq('ACAACCTAGGTGTACCTAGTTAGTGGGTTA')]),\n",
       "             ('test_seq_22', [Seq('ACAACTCCTCTACTTCTAAAGTCTGGGTTA')]),\n",
       "             ('test_seq_23', [Seq('ACAAAGCTACCTAGCCTCCCTACTGGGTTA')]),\n",
       "             ('test_seq_24', [Seq('ACAACTAGCCTAGTCTAGTTACGAGGGTTA')]),\n",
       "             ('test_seq_25', [Seq('ACAATAGACCTAGATAGTTTCCAAGGGTTA')]),\n",
       "             ('test_seq_26', [Seq('ACAAGACCCCCCTCAGCTACCTAGGGGTTA')]),\n",
       "             ('test_seq_27', [Seq('ACAATAGTTTCCAAGACCCCCCTCGGGTTA')]),\n",
       "             ('test_seq_28', [Seq('ACAAAGCTACCTAGTTAGTGAGACGGGTTA')]),\n",
       "             ('test_seq_29', [Seq('ACAAGACTCACGAGGTCTATAGTAGGGTTA')]),\n",
       "             ('test_seq_30', [Seq('ACAACTAGAAGCTATCTAAAGTCTGGGTTA')]),\n",
       "             ('test_seq_31', [Seq('ACAAATGTAACTAGCTCCTCTACTGGGTTA')]),\n",
       "             ('test_seq_32', [Seq('ACAACCTAGGACCCCCTAGGACCCGGGTTA')]),\n",
       "             ('test_seq_33', [Seq('ACAACTAGTTACGAATAGTACTAGGGGTTA')]),\n",
       "             ('test_seq_34', [Seq('ACAATAGTTTCCAATTAGTGAGACGGGTTA')]),\n",
       "             ('test_seq_35', [Seq('ACAACTAGTTACGAGACCCCCCTCGGGTTA')]),\n",
       "             ('test_seq_36', [Seq('ACAAGACTCACGAGTAGACCTAGAGGGTTA')]),\n",
       "             ('test_seq_37', [Seq('ACAATAGACTAGAATAGACTAGAAGGGTTA')]),\n",
       "             ('test_seq_38', [Seq('ACAATCTAAAGTCTTAGACCTAGAGGGTTA')]),\n",
       "             ('test_seq_39', [Seq('ACAACTAGAGTCTCCTAGTTACGAGGGTTA')]),\n",
       "             ('test_seq_40', [Seq('ACAATAGACTAGAACTAGAAGCTAGGGTTA')]),\n",
       "             ('test_seq_41', [Seq('ACAAGTCTATAGTACTAGAAGCTAGGGTTA')]),\n",
       "             ('test_seq_42', [Seq('ACAACTAGTTTAGAATGTAACTAGGGGTTA')]),\n",
       "             ('test_seq_43', [Seq('ACAAATAGTACTAGCCTCCCTACTGGGTTA')]),\n",
       "             ('test_seq_44', [Seq('ACAATTAGTGAGACTAGACCTAGAGGGTTA')]),\n",
       "             ('test_seq_45', [Seq('ACAATCTAAAGTCTGACCCCCCTCGGGTTA')]),\n",
       "             ('test_seq_46', [Seq('ACAAACTAGCTAGCACTAGCTAGCGGGTTA')]),\n",
       "             ('test_seq_47', [Seq('ACAATTAGTGAGACAGCTACCTAGGGGTTA')]),\n",
       "             ('test_seq_48', [Seq('ACAAGACCCCCCTCCCTAGGACCCGGGTTA')]),\n",
       "             ('test_seq_49', [Seq('ACAAACTAGCTAGCCCTAGGTGTAGGGTTA')]),\n",
       "             ('test_seq_50', [Seq('ACAACCTAGGACCCATGTAACTAGGGGTTA')])])"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_data_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WARNING:tensorflow:SavedModel saved prior to TF 2.5 detected when loading Keras model. Please ensure that you are saving the model with model.save() or tf.keras.models.save_model(), *NOT* tf.saved_model.save(). To confirm, there should be a file named \"keras_metadata.pb\" in the SavedModel directory.\n",
      "2/2 [==============================] - 1s 215ms/step\n",
      "WARNING:tensorflow:SavedModel saved prior to TF 2.5 detected when loading Keras model. Please ensure that you are saving the model with model.save() or tf.keras.models.save_model(), *NOT* tf.saved_model.save(). To confirm, there should be a file named \"keras_metadata.pb\" in the SavedModel directory.\n",
      "2/2 [==============================] - 1s 267ms/step\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(             ID                                    30mer_sequences  On_Scores\n",
       " 0    test_seq_1  (A, C, A, A, A, T, G, T, A, A, C, T, A, G, T, ...  28.877352\n",
       " 1    test_seq_2  (A, C, A, A, T, T, A, G, T, G, A, G, A, C, C, ...  55.356281\n",
       " 2    test_seq_3  (A, C, A, A, C, C, T, A, G, T, T, A, G, T, A, ...  46.279396\n",
       " 3    test_seq_4  (A, C, A, A, G, T, C, C, T, T, C, T, A, G, C, ...  24.130835\n",
       " 4    test_seq_5  (A, C, A, A, C, C, T, A, G, G, T, G, T, A, T, ...  53.853676\n",
       " 5    test_seq_6  (A, C, A, A, C, T, A, G, T, T, T, A, G, A, G, ...  60.841446\n",
       " 6    test_seq_7  (A, C, A, A, A, C, C, T, A, A, G, T, A, T, T, ...  21.330648\n",
       " 7    test_seq_8  (A, C, A, A, A, G, C, T, A, C, C, T, A, G, G, ...  65.294968\n",
       " 8    test_seq_9  (A, C, A, A, G, A, C, C, C, C, C, C, T, C, C, ...  24.737684\n",
       " 9   test_seq_10  (A, C, A, A, C, C, T, A, G, G, A, C, C, C, C, ...  16.199524\n",
       " 10  test_seq_11  (A, C, A, A, G, A, C, C, C, C, C, C, T, C, G, ...  27.935749\n",
       " 11  test_seq_12  (A, C, A, A, G, T, C, T, A, T, A, G, T, A, G, ...  39.898598\n",
       " 12  test_seq_13  (A, C, A, A, C, C, T, C, C, C, T, A, C, T, T, ...  46.381046\n",
       " 13  test_seq_14  (A, C, A, A, C, C, T, C, C, C, T, A, C, T, A, ...  44.405281\n",
       " 14  test_seq_15  (A, C, A, A, G, A, C, C, C, C, C, C, T, C, A, ...  67.924210\n",
       " 15  test_seq_16  (A, C, A, A, T, A, G, A, C, C, T, A, G, A, C, ...  43.503536\n",
       " 16  test_seq_17  (A, C, A, A, T, C, T, A, A, A, G, T, C, T, C, ...  24.642365\n",
       " 17  test_seq_18  (A, C, A, A, T, C, T, A, A, A, G, T, C, T, C, ...  58.648239\n",
       " 18  test_seq_19  (A, C, A, A, T, C, T, A, A, A, G, T, C, T, G, ...  45.078156\n",
       " 19  test_seq_20  (A, C, A, A, C, A, A, G, A, A, G, T, A, T, G, ...  45.285393\n",
       " 20  test_seq_21  (A, C, A, A, C, C, T, A, G, G, T, G, T, A, C, ...  22.499449\n",
       " 21  test_seq_22  (A, C, A, A, C, T, C, C, T, C, T, A, C, T, T, ...  43.876644\n",
       " 22  test_seq_23  (A, C, A, A, A, G, C, T, A, C, C, T, A, G, C, ...  34.138599\n",
       " 23  test_seq_24  (A, C, A, A, C, T, A, G, C, C, T, A, G, T, C, ...  58.095284\n",
       " 24  test_seq_25  (A, C, A, A, T, A, G, A, C, C, T, A, G, A, T, ...  59.994591\n",
       " 25  test_seq_26  (A, C, A, A, G, A, C, C, C, C, C, C, T, C, A, ...  67.924210\n",
       " 26  test_seq_27  (A, C, A, A, T, A, G, T, T, T, C, C, A, A, G, ...  47.439262\n",
       " 27  test_seq_28  (A, C, A, A, A, G, C, T, A, C, C, T, A, G, T, ...  25.374310\n",
       " 28  test_seq_29  (A, C, A, A, G, A, C, T, C, A, C, G, A, G, G, ...  52.363472\n",
       " 29  test_seq_30  (A, C, A, A, C, T, A, G, A, A, G, C, T, A, T, ...  45.242783\n",
       " 30  test_seq_31  (A, C, A, A, A, T, G, T, A, A, C, T, A, G, C, ...  39.061089\n",
       " 31  test_seq_32  (A, C, A, A, C, C, T, A, G, G, A, C, C, C, C, ...  44.886513\n",
       " 32  test_seq_33  (A, C, A, A, C, T, A, G, T, T, A, C, G, A, A, ...  47.005558\n",
       " 33  test_seq_34  (A, C, A, A, T, A, G, T, T, T, C, C, A, A, T, ...  25.521549\n",
       " 34  test_seq_35  (A, C, A, A, C, T, A, G, T, T, A, C, G, A, G, ...  40.077240\n",
       " 35  test_seq_36  (A, C, A, A, G, A, C, T, C, A, C, G, A, G, T, ...  50.387352\n",
       " 36  test_seq_37  (A, C, A, A, T, A, G, A, C, T, A, G, A, A, T, ...  52.747425\n",
       " 37  test_seq_38  (A, C, A, A, T, C, T, A, A, A, G, T, C, T, T, ...  27.954227\n",
       " 38  test_seq_39  (A, C, A, A, C, T, A, G, A, G, T, C, T, C, C, ...  60.279839\n",
       " 39  test_seq_40  (A, C, A, A, T, A, G, A, C, T, A, G, A, A, C, ...  72.846046\n",
       " 40  test_seq_41  (A, C, A, A, G, T, C, T, A, T, A, G, T, A, C, ...  61.723999\n",
       " 41  test_seq_42  (A, C, A, A, C, T, A, G, T, T, T, A, G, A, A, ...  47.122932\n",
       " 42  test_seq_43  (A, C, A, A, A, T, A, G, T, A, C, T, A, G, C, ...  36.144211\n",
       " 43  test_seq_44  (A, C, A, A, T, T, A, G, T, G, A, G, A, C, T, ...  47.271297\n",
       " 44  test_seq_45  (A, C, A, A, T, C, T, A, A, A, G, T, C, T, G, ...  45.205662\n",
       " 45  test_seq_46  (A, C, A, A, A, C, T, A, G, C, T, A, G, C, A, ...  24.245186\n",
       " 46  test_seq_47  (A, C, A, A, T, T, A, G, T, G, A, G, A, C, A, ...  66.729553\n",
       " 47  test_seq_48  (A, C, A, A, G, A, C, C, C, C, C, C, T, C, C, ...  57.359116\n",
       " 48  test_seq_49  (A, C, A, A, A, C, T, A, G, C, T, A, G, C, C, ...  53.381580\n",
       " 49  test_seq_50  (A, C, A, A, C, C, T, A, G, G, A, C, C, C, A, ...  59.612793,\n",
       "              ID                                    30mer_sequences  On_Scores\n",
       " 0    test_seq_1  (A, G, A, T, A, T, G, T, A, A, C, T, A, G, T, ...  28.709793\n",
       " 1    test_seq_2  (A, G, A, T, T, T, A, G, T, G, A, G, A, C, C, ...  55.711227\n",
       " 2    test_seq_3  (A, G, A, T, C, C, T, A, G, T, T, A, G, T, A, ...  49.223003\n",
       " 3    test_seq_4  (A, G, A, T, G, T, C, C, T, T, C, T, A, G, C, ...  25.802984\n",
       " 4    test_seq_5  (A, G, A, T, C, C, T, A, G, G, T, G, T, A, T, ...  56.032192\n",
       " 5    test_seq_6  (A, G, A, T, C, T, A, G, T, T, T, A, G, A, G, ...  61.907669\n",
       " 6    test_seq_7  (A, G, A, T, A, C, C, T, A, A, G, T, A, T, T, ...  20.388386\n",
       " 7    test_seq_8  (A, G, A, T, A, G, C, T, A, C, C, T, A, G, G, ...  69.053986\n",
       " 8    test_seq_9  (A, G, A, T, G, A, C, C, C, C, C, C, T, C, C, ...  25.694763\n",
       " 9   test_seq_10  (A, G, A, T, C, C, T, A, G, G, A, C, C, C, C, ...  17.359648\n",
       " 10  test_seq_11  (A, G, A, T, G, A, C, C, C, C, C, C, T, C, G, ...  28.656069\n",
       " 11  test_seq_12  (A, G, A, T, G, T, C, T, A, T, A, G, T, A, G, ...  41.574146\n",
       " 12  test_seq_13  (A, G, A, T, C, C, T, C, C, C, T, A, C, T, T, ...  47.437504\n",
       " 13  test_seq_14  (A, G, A, T, C, C, T, C, C, C, T, A, C, T, A, ...  48.264099\n",
       " 14  test_seq_15  (A, G, A, T, G, A, C, C, C, C, C, C, T, C, A, ...  70.157036\n",
       " 15  test_seq_16  (A, G, A, T, T, A, G, A, C, C, T, A, G, A, C, ...  42.074875\n",
       " 16  test_seq_17  (A, G, A, T, T, C, T, A, A, A, G, T, C, T, C, ...  23.359934\n",
       " 17  test_seq_18  (A, G, A, T, T, C, T, A, A, A, G, T, C, T, C, ...  58.651703\n",
       " 18  test_seq_19  (A, G, A, T, T, C, T, A, A, A, G, T, C, T, G, ...  45.513023\n",
       " 19  test_seq_20  (A, G, A, T, C, A, A, G, A, A, G, T, A, T, G, ...  47.203762\n",
       " 20  test_seq_21  (A, G, A, T, C, C, T, A, G, G, T, G, T, A, C, ...  23.866615\n",
       " 21  test_seq_22  (A, G, A, T, C, T, C, C, T, C, T, A, C, T, T, ...  43.239193\n",
       " 22  test_seq_23  (A, G, A, T, A, G, C, T, A, C, C, T, A, G, C, ...  33.568279\n",
       " 23  test_seq_24  (A, G, A, T, C, T, A, G, C, C, T, A, G, T, C, ...  58.136280\n",
       " 24  test_seq_25  (A, G, A, T, T, A, G, A, C, C, T, A, G, A, T, ...  57.979347\n",
       " 25  test_seq_26  (A, G, A, T, G, A, C, C, C, C, C, C, T, C, A, ...  70.157036\n",
       " 26  test_seq_27  (A, G, A, T, T, A, G, T, T, T, C, C, A, A, G, ...  46.698093\n",
       " 27  test_seq_28  (A, G, A, T, A, G, C, T, A, C, C, T, A, G, T, ...  22.707455\n",
       " 28  test_seq_29  (A, G, A, T, G, A, C, T, C, A, C, G, A, G, G, ...  53.288063\n",
       " 29  test_seq_30  (A, G, A, T, C, T, A, G, A, A, G, C, T, A, T, ...  46.786293\n",
       " 30  test_seq_31  (A, G, A, T, A, T, G, T, A, A, C, T, A, G, C, ...  40.405678\n",
       " 31  test_seq_32  (A, G, A, T, C, C, T, A, G, G, A, C, C, C, C, ...  48.273235\n",
       " 32  test_seq_33  (A, G, A, T, C, T, A, G, T, T, A, C, G, A, A, ...  47.182968\n",
       " 33  test_seq_34  (A, G, A, T, T, A, G, T, T, T, C, C, A, A, T, ...  23.175301\n",
       " 34  test_seq_35  (A, G, A, T, C, T, A, G, T, T, A, C, G, A, G, ...  44.666664\n",
       " 35  test_seq_36  (A, G, A, T, G, A, C, T, C, A, C, G, A, G, T, ...  51.025124\n",
       " 36  test_seq_37  (A, G, A, T, T, A, G, A, C, T, A, G, A, A, T, ...  50.664005\n",
       " 37  test_seq_38  (A, G, A, T, T, C, T, A, A, A, G, T, C, T, T, ...  25.775887\n",
       " 38  test_seq_39  (A, G, A, T, C, T, A, G, A, G, T, C, T, C, C, ...  58.845970\n",
       " 39  test_seq_40  (A, G, A, T, T, A, G, A, C, T, A, G, A, A, C, ...  72.380463\n",
       " 40  test_seq_41  (A, G, A, T, G, T, C, T, A, T, A, G, T, A, C, ...  62.976555\n",
       " 41  test_seq_42  (A, G, A, T, C, T, A, G, T, T, T, A, G, A, A, ...  46.713623\n",
       " 42  test_seq_43  (A, G, A, T, A, T, A, G, T, A, C, T, A, G, C, ...  38.852966\n",
       " 43  test_seq_44  (A, G, A, T, T, T, A, G, T, G, A, G, A, C, T, ...  45.025913\n",
       " 44  test_seq_45  (A, G, A, T, T, C, T, A, A, A, G, T, C, T, G, ...  44.208019\n",
       " 45  test_seq_46  (A, G, A, T, A, C, T, A, G, C, T, A, G, C, A, ...  26.988909\n",
       " 46  test_seq_47  (A, G, A, T, T, T, A, G, T, G, A, G, A, C, A, ...  67.364655\n",
       " 47  test_seq_48  (A, G, A, T, G, A, C, C, C, C, C, C, T, C, C, ...  60.352077\n",
       " 48  test_seq_49  (A, G, A, T, A, C, T, A, G, C, T, A, G, C, C, ...  56.655125\n",
       " 49  test_seq_50  (A, G, A, T, C, C, T, A, G, G, A, C, C, C, A, ...  62.808331)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1=CRISPROn_score_pipeline(seq_data_1)\n",
    "result_2=CRISPROn_score_pipeline(seq_data_2)\n",
    "result_1, result_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_1.to_csv('./result_1.csv',index=False)\n",
    "result_2.to_csv('./result_2.csv',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "[1] CRISPRon: enhanced data-driven on-target CRISPR-Cas9 gRNA efficiency prediction. Xiang X, Corsi GI, Anthon C, Qu K, , Pan X, Liang X, Han P, Dong Z, Liu L, Zhong J, Ma T, Wang J, Zhang X, Jiang H, Xu F, Liu X, Xu X, Wang J, Yang H, Bolund L, Church GM, Lin L, Gorodkin J, Luo Y. Submitted.\n",
    "\n",
    "[2] SpCas9 activity prediction by DeepSpCas9, a deep learning-based model with high generalization performance. Kim, H.K. et al. Sci Adv 5, eaax9249 (2019).\n",
    "\n",
    "[3] CRISPR-Cas9 off-targeting assessment with nucleic acid duplex energy parameters. Alkan F, Wenzel A, Anthon C, Havgaard JH, Gorodkin J Genome Biol. 2018 Oct 26;19(1):177"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.15"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
